// rclnodejs/generated/jackal_msgs/jackal_msgs__msg__Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const DriveFeedbackWrapper = require('../../generated/jackal_msgs/jackal_msgs__msg__DriveFeedback.js');
const FeedbackRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  drivers: ArrayType(DriveFeedbackWrapper.refObjectType, 2),
  commanded_mode: primitiveTypes.int8,
  actual_mode: primitiveTypes.int8,
});
const FeedbackRefArray = ArrayType(FeedbackRefStruct);
const FeedbackRefStructArray = StructType({
  data: FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._wrapperFields.drivers = DriveFeedbackWrapper.createArray();
      this._wrapperFields.drivers.copy(other._wrapperFields.drivers);
      this._commanded_modeIntialized = true;
      this._actual_modeIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new FeedbackRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._wrapperFields.drivers = DriveFeedbackWrapper.createArray();
    this._commanded_modeIntialized = false;
    this._actual_modeIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return FeedbackRefStructArray
  }
  static get refObjectType() {
    return FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._commanded_modeIntialized) {
        throw new TypeError('Invalid argument: commanded_mode in Feedback');
      }
      if (!this._actual_modeIntialized) {
        throw new TypeError('Invalid argument: actual_mode in Feedback');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    for (let i = 0; i < this._wrapperFields.drivers.data.length; i++) {
      this._refObject.drivers[i] = this._wrapperFields.drivers.data[i].freeze(own, checkConsistency);
      this._refObject.drivers[i] = this._wrapperFields.drivers.data[i].refObject;
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._refObject.drivers = refObject.drivers;
    this._wrapperFields.drivers.size = 2
    for (let i = 0; i < 2; i++) {
      this._wrapperFields.drivers.data[i].copyRefObject(refObject.drivers[i]);
    }
    this._commanded_modeIntialized = true;
    this._refObject.commanded_mode = refObject.commanded_mode;
    this._actual_modeIntialized = true;
    this._refObject.actual_mode = refObject.actual_mode;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    for (let i = 0; i < 2; i++) {
      DriveFeedbackWrapper.freeStruct(refObject.drivers[i]);
    }
  }
  static destoryRawROS(msg) {
    FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'jackal_msgs', subFolder: 'msg', interfaceName: 'Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get drivers() {
    return this._wrapperFields.drivers;
  }
  set drivers(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields.drivers.fill(value);
  }
  get commanded_mode() {
    if (!this._commanded_modeIntialized) {
      return undefined;
    }
    return this._refObject.commanded_mode;
  }
  set commanded_mode(value) {
    this._commanded_modeIntialized = true;
    this._refObject.commanded_mode = value;
  }
  get actual_mode() {
    if (!this._actual_modeIntialized) {
      return undefined;
    }
    return this._refObject.actual_mode;
  }
  set actual_mode(value) {
    this._actual_modeIntialized = true;
    this._refObject.actual_mode = value;
  }
  copyRefObject(refObject) {
    this._refObject = new FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._refObject.drivers = refObject.drivers;
    this._wrapperFields.drivers.size = 2
    for (let i = 0; i < 2; i++) {
      this._wrapperFields.drivers.data[i].copyRefObject(refObject.drivers[i]);
    }
    this._commanded_modeIntialized = true;
    this._actual_modeIntialized = true;
  }
  copy(other) {
    this._refObject = new FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._wrapperFields.drivers.copy(other._wrapperFields.drivers);
    this._commanded_modeIntialized = true;
    this._actual_modeIntialized = true;
  }
  static get classType() {
    return FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"drivers","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":"jackal_msgs","type":"DriveFeedback","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"commanded_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"actual_mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"jackal_msgs","type":"Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Feedback"};
  }
  hasMember(name) {
    let memberNames = ["header","drivers","commanded_mode","actual_mode"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new FeedbackRefArray(size);
    this._refObject = new FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return FeedbackArrayWrapper;
  }
}
module.exports = FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "drivers",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": "jackal_msgs",
        "type": "DriveFeedback",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "commanded_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "actual_mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "jackal_msgs",
    "type": "Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Feedback"
}
*/
