// rclnodejs/generated/jackal_msgs/jackal_msgs__msg__Drive.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const Float32Wrapper = require('../../generated/std_msgs/std_msgs__msg__Float32.js');
const DriveRefStruct = StructType({
  mode: primitiveTypes.int8,
  drivers: ArrayType(primitiveTypes.float32, 2),
});
const DriveRefArray = ArrayType(DriveRefStruct);
const DriveRefStructArray = StructType({
  data: DriveRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class DriveWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new DriveRefStruct(other._refObject.toObject());
      this._modeIntialized = true;
      this._wrapperFields.drivers = Float32Wrapper.createArray();
      this._wrapperFields.drivers.copy(other._wrapperFields.drivers);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new DriveRefStruct();
    this._modeIntialized = false;
    this._wrapperFields.drivers = Float32Wrapper.createArray();
  }
  static createFromRefObject(refObject) {
    let self = new DriveWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new DriveArrayWrapper;
  }
  static get ArrayType() {
    return DriveArrayWrapper;
  }
  static get refObjectArrayType() {
    return DriveRefStructArray
  }
  static get refObjectType() {
    return DriveRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._modeIntialized) {
        throw new TypeError('Invalid argument: mode in Drive');
      }
    }
    for (let i = 0; i < this._wrapperFields.drivers.data.length; i++) {
      this._refObject.drivers[i] = this._wrapperFields.drivers.data[i];
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._modeIntialized = true;
    this._refObject.mode = refObject.mode;
    this._refObject.drivers = refObject.drivers;
    this._wrapperFields.drivers.fill(refObject.drivers.toArray());
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    DriveWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'jackal_msgs', subFolder: 'msg', interfaceName: 'Drive'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get mode() {
    if (!this._modeIntialized) {
      return undefined;
    }
    return this._refObject.mode;
  }
  set mode(value) {
    this._modeIntialized = true;
    this._refObject.mode = value;
  }
  get drivers() {
    return this._wrapperFields['drivers'].data;
  }
  set drivers(value) {
    if (value.length !== 2) {
      throw new RangeError('The length of the array must be 2.');
    }
    this._wrapperFields['drivers'].fill(value);
  }
  copyRefObject(refObject) {
    this._refObject = new DriveRefStruct(refObject.toObject());
    this._modeIntialized = true;
    this._wrapperFields.drivers.fill(refObject.drivers.toArray());
  }
  copy(other) {
    this._refObject = new DriveRefStruct(other._refObject.toObject());
    this._modeIntialized = true;
    this._wrapperFields.drivers.copy(other._wrapperFields.drivers);
  }
  static get classType() {
    return DriveWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[{"type":"int8","name":"MODE_VELOCITY","value":0},{"type":"int8","name":"MODE_PWM","value":1},{"type":"int8","name":"MODE_EFFORT","value":2},{"type":"int8","name":"MODE_NONE","value":-1},{"type":"int8","name":"LEFT","value":0},{"type":"int8","name":"RIGHT","value":1}],"fields":[{"name":"mode","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"drivers","type":{"isArray":true,"arraySize":2,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":true,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"jackal_msgs","type":"Drive","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Drive"};
  }
  hasMember(name) {
    let memberNames = ["mode","drivers"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class DriveArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof DriveWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new DriveWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DriveArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to DriveArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new DriveRefArray(size);
    this._refObject = new DriveRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new DriveWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof DriveArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "DriveArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      DriveWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return DriveWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return DriveArrayWrapper;
  }
}
// Define constants (6 in total)
Object.defineProperty(DriveWrapper, "MODE_VELOCITY", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DriveWrapper, "MODE_PWM", {value: 1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DriveWrapper, "MODE_EFFORT", {value: 2, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DriveWrapper, "MODE_NONE", {value: -1, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DriveWrapper, "LEFT", {value: 0, writable: false, enumerable: true, configurable: true});
Object.defineProperty(DriveWrapper, "RIGHT", {value: 1, writable: false, enumerable: true, configurable: true});
module.exports = DriveWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [
    {
      "type": "int8",
      "name": "MODE_VELOCITY",
      "value": 0
    },
    {
      "type": "int8",
      "name": "MODE_PWM",
      "value": 1
    },
    {
      "type": "int8",
      "name": "MODE_EFFORT",
      "value": 2
    },
    {
      "type": "int8",
      "name": "MODE_NONE",
      "value": -1
    },
    {
      "type": "int8",
      "name": "LEFT",
      "value": 0
    },
    {
      "type": "int8",
      "name": "RIGHT",
      "value": 1
    }
  ],
  "fields": [
    {
      "name": "mode",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "drivers",
      "type": {
        "isArray": true,
        "arraySize": 2,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": true,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "jackal_msgs",
    "type": "Drive",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Drive"
}
*/
